drivers/pinctrl/pinctrl_gd32_afio.c

```diff
@@ -0,0 +1,210 @@
+	GD32_PORT_ADDR_OR_NONE(gpiog)
+};
+
+/** GD32 port RCUs */
+static const uint32_t gd32_port_rcus[] = {
+	GD32_PORT_RCU_OR_NONE(gpioa)
+	GD32_PORT_RCU_OR_NONE(gpiob)
+	GD32_PORT_RCU_OR_NONE(gpioc)
+	GD32_PORT_RCU_OR_NONE(gpiod)
+	GD32_PORT_RCU_OR_NONE(gpioe)
+	GD32_PORT_RCU_OR_NONE(gpiof)
+	GD32_PORT_RCU_OR_NONE(gpiog)
+};
+
+/**
+ * @brief Initialize AFIO
+ *
+ * This function enables AFIO clock and configures the I/O compensation if
+ * available and enabled in Devicetree.
+ *
+ * @retval 0 Always
+ */
+static int afio_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	rcu_periph_clock_enable(DT_PROP(AFIO_NODE, rcu_periph_clock));
+
+#ifdef AFIO_CPSCTL
+	if (DT_PROP(AFIO_NODE, enable_cps)) {
+		gpio_compensation_config(GPIO_COMPENSATION_ENABLE);
+		while (gpio_compensation_flag_get() == RESET)
+			;
+	}
+#endif /* AFIO_CPSCTL */
+
+	return 0;
+}
+
+SYS_INIT(afio_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+/**
+ * @brief Helper function to configure the SPD register if available.
+ *
+ * @param port GPIO port address.
+ * @param pin_num GPIO pin.
+ * @param speed GPIO speed.
+ *
+ * @return Value of the mode register that should be set later.
+ */
+static inline uint8_t configure_spd(uint32_t port, uint8_t pin_num,
+				    uint8_t speed)
+{
+	switch (speed) {
+	case GD32_OSPEED_MAX:
+#ifdef GPIOx_SPD
+		GPIOx_SPD(port) |= BIT(pin_num);
+#endif /* GPIOx_SPD */
+		return speed;
+	default:
+#ifdef GPIOx_SPD
+		GPIOx_SPD(port) &= ~BIT(pin_num);
+#endif /* GPIOx_SPD */
+		return speed + 1U;
+	}
+}
+
+/**
+ * @brief Configure a pin.
+ *
+ * @param pin The pin to configure.
+ */
+static void configure_pin(pinctrl_soc_pin_t pin)
+{
+	uint8_t port_idx, mode, pin_num;
+	uint32_t rcu, port;
+
+	port_idx = GD32_PORT_GET(pin);
+	__ASSERT_NO_MSG(port_idx < ARRAY_SIZE(gd32_port_addrs));
+
+	rcu = gd32_port_rcus[port_idx];
+	port = gd32_port_addrs[port_idx];
+	pin_num = GD32_PIN_GET(pin);
+	mode = GD32_MODE_GET(pin);
+
+	rcu_periph_clock_enable(rcu);
+
+	if (pin_num < 8U) {
+		uint32_t reg = GPIO_CTL0(port);
+
+		reg &= ~GPIO_MODE_MASK(pin_num);
+
+		/* configure output speed */
+		if (mode == GD32_MODE_ALTERNATE) {
+			uint8_t speed = GD32_OSPEED_GET(pin);
+
+			speed = configure_spd(port, pin_num, speed);
+			reg |= GPIO_MODE_SET(pin_num, speed);
+		}
+
+		GPIO_CTL0(port) = reg;
+	} else {
+		uint32_t reg = GPIO_CTL1(port);
+
+		reg &= ~GPIO_MODE_MASK(pin_num - 8U);
+
+		/* configure output speed */
+		if (mode == GD32_MODE_ALTERNATE) {
+			uint8_t speed = GD32_OSPEED_GET(pin);
+
+			speed = configure_spd(port, pin_num, speed);
+			reg |= GPIO_MODE_SET(pin_num - 8U, speed);
+		}
+
+		GPIO_CTL1(port) = reg;
+	}
+
+	/* configure pull-up/down if input */
+	if (mode == GD32_MODE_GPIO_IN) {
+		uint8_t pupd = GD32_PUPD_GET(pin);
+
+		if (pupd == GD32_PUPD_PULLDOWN) {
+			GPIO_BC(port) = BIT(pin_num);
+		} else if (pupd == GD32_PUPD_PULLUP) {
+			GPIO_BOP(port) = BIT(pin_num);
+		}
+	}
+
+	rcu_periph_clock_disable(rcu);
+}
```

Comment:
Hi @gmarull ,

The code as is not select Output AFIO mode at CTLx[1:0], where x is the pin_num. I made some changes based in the below table. After that, USART start to work. I noted that input-floating (bias-high-impedance) is missing.
My feeling is that `configure_pin` function should be reviewed to match with GPIO configuration table.

![GPIO_configuration_table](https://user-images.githubusercontent.com/7027887/141598514-a293aec5-25de-43eb-b150-22979bcc42b5.jpg)

```diff
git diff drivers/pinctrl/pinctrl_gd32_afio.c
diff --git a/drivers/pinctrl/pinctrl_gd32_afio.c b/drivers/pinctrl/pinctrl_gd32_afio.c
index 613e065d8c..a3ad522e6c 100644
--- a/drivers/pinctrl/pinctrl_gd32_afio.c
+++ b/drivers/pinctrl/pinctrl_gd32_afio.c
@@ -102,7 +102,8 @@ static inline uint8_t configure_spd(uint32_t port, uint8_t pin_num,
 static void configure_pin(pinctrl_soc_pin_t pin)
 {
        uint8_t port_idx, mode, pin_num;
-       uint32_t rcu, port;
+       uint32_t rcu, port, reg_tmp;
+       volatile uint32_t *reg;
 
        port_idx = GD32_PORT_GET(pin);
        __ASSERT_NO_MSG(port_idx < ARRAY_SIZE(gd32_port_addrs));
@@ -111,50 +112,35 @@ static void configure_pin(pinctrl_soc_pin_t pin)
        port = gd32_port_addrs[port_idx];
        pin_num = GD32_PIN_GET(pin);
        mode = GD32_MODE_GET(pin);
+       reg = (pin_num < 8U) ? &GPIO_CTL0(port) : &GPIO_CTL1(port);
+       pin_num = (pin_num < 8U) ? pin_num : (pin_num - 8U);
 
        rcu_periph_clock_enable(rcu);
 
-       if (pin_num < 8U) {
-               uint32_t reg = GPIO_CTL0(port);
+       reg_tmp = *reg;
+       reg_tmp &= ~GPIO_MODE_MASK(pin_num);
 
-               reg &= ~GPIO_MODE_MASK(pin_num);
+       if (mode == GD32_MODE_ALTERNATE) {
+               uint8_t speed = GD32_OSPEED_GET(pin);
 
-               /* configure output speed */
-               if (mode == GD32_MODE_ALTERNATE) {
-                       uint8_t speed = GD32_OSPEED_GET(pin);
+               speed = configure_spd(port, pin_num, speed);
+               reg_tmp |= GPIO_MODE_SET(pin_num, speed | 0x08);
 
-                       speed = configure_spd(port, pin_num, speed);
-                       reg |= GPIO_MODE_SET(pin_num, speed);
-               }
-
-               GPIO_CTL0(port) = reg;
-       } else {
-               uint32_t reg = GPIO_CTL1(port);
-
-               reg &= ~GPIO_MODE_MASK(pin_num - 8U);
-
-               /* configure output speed */
-               if (mode == GD32_MODE_ALTERNATE) {
-                       uint8_t speed = GD32_OSPEED_GET(pin);
-
-                       speed = configure_spd(port, pin_num, speed);
-                       reg |= GPIO_MODE_SET(pin_num - 8U, speed);
-               }
-
-               GPIO_CTL1(port) = reg;
-       }
-
-       /* configure pull-up/down if input */
-       if (mode == GD32_MODE_GPIO_IN) {
+               /* TODO: OPEN-DRAIN */
+       } else if (mode == GD32_MODE_GPIO_IN) {
                uint8_t pupd = GD32_PUPD_GET(pin);
 
                if (pupd == GD32_PUPD_PULLDOWN) {
                        GPIO_BC(port) = BIT(pin_num);
+                       reg_tmp |= GPIO_MODE_SET(pin_num, 0x08); /* PUPD Mode */
                } else if (pupd == GD32_PUPD_PULLUP) {
                        GPIO_BOP(port) = BIT(pin_num);
+                       reg_tmp |= GPIO_MODE_SET(pin_num, 0x08); /* PUPD Mode */
                }
        }
 
+       *reg = reg_tmp;
+
        rcu_periph_clock_disable(rcu);
 }
```

Reply:
Thanks for the testing and suggestions, seems like I missed quite a few things. I have updated the driver, I hope all things are covered now.